packages/drawer/src/drawer.vue

```diff
@@ -0,0 +1,238 @@
+      type: Boolean,
+      default: false,
+    },
+    beforeClose: {
+      type: Function as PropType<(any) => void>,
+      default: null,
+    },
+    customClass: {
+      type: String,
+      default: '',
+    },
+    closeOnPressEscape: {
+      type: Boolean,
+      default: true,
+    },
+    destroyOnClose: {
+      type: Boolean,
+      default: false,
+    },
+    modal: {
+      type: Boolean,
+      default: true,
+    },
+    direction: {
+      type: String as PropType<IDrawerDirection>,
+      default: 'rtl',
+      validator(val: IDrawerDirection) {
+        return ['ltr', 'rtl', 'ttb', 'btt'].indexOf(val) !== -1
+      },
+    },
+    modalAppendToBody: {
+      type: Boolean,
+      default: true,
+    },
+    showClose: {
+      type: Boolean,
+      default: true,
+    },
+    size: {
+      type: String,
+      default: '30%',
+    },
+    title: {
+      type: String,
+      default: '',
+    },
+    visible: {
+      type: Boolean,
+    },
+    wrapperClosable: {
+      type: Boolean,
+      default: true,
+    },
+    withHeader: {
+      type: Boolean,
+      default: true,
+    },
+  },
+
+  emits: ['open', 'opened', 'close', 'closed', 'update:visible'],
+
+  setup(props: IElDrawerProps, ctx) {
+    const drawer = ref<HTMLElement>(null)
+    const root = ref<HTMLElement>(null)
+    const prevActiveElement = ref<HTMLElement>(null)
+    const closed = ref(false)
+    const isHorizontal = computed(
+      () => props.direction === 'rtl' || props.direction === 'ltr',
+    )
+
+    const { popupData, open } = usePopup({
+      props,
+      $el: root,
+      vVm: { handleClose, closeOnClickModal: props.closeOnClickModal },
+    })
+
+    function afterEnter() {
+      ctx.emit('opened')
+    }
+
+    function afterLeave() {
+      ctx.emit('closed')
+    }
+
+    function hide(cancel = true) {
+      if (cancel !== false) {
+        ctx.emit('update:visible', false)
+        ctx.emit('close')
+        if (props.destroyOnClose === true) {
+          popupData.rendered = false
+        }
+        closed.value = true
+      }
+    }
+
+    function handleWrapperClick() {
+      if (props.wrapperClosable) {
+        closeDrawer()
+      }
+    }
+
+    function closeDrawer() {
+      if (typeof props.beforeClose === 'function') {
+        props.beforeClose(hide)
+      } else {
+        hide()
+      }
+    }
+
+    function handleClose() {
+      // This method here will be called by PopupManger, when the `closeOnPressEscape` was set to true
+      // pressing `ESC` will call this method, and also close the drawer.
+      // This method also calls `beforeClose` if there was one.
+      closeDrawer()
+    }
+
+    watch(
+      () => props.visible,
+      val => {
+        if (val) {
+          closed.value = false
+          ctx.emit('open')
+          // if (props.appendToBody) {
+          //   document.body.appendChild(root.value)
+          // }
+          prevActiveElement.value = document.activeElement as HTMLElement
```

Comment:
I actually had a composition hook in #197, but lets stick with this for now, I'll do the optimizations after this gets merged